components/fraud_proof.md

```diff
@@ -0,0 +1,64 @@
+# Fraud Proof Algorithm for Simple Sequencer
+This is a WIP -- mostly just a few notes that I'm jotting down to not forget.
+
+## Requirements
+1. Not all blockhashes are posted up front.
+    - Blockhashes should be posted roughly every 100 L1 blocks.
+2. Only one block is evaluated during the fraud proof.
+
+## Fraud Proof Structure
+
+Fraud proofs are initialized with some constant size input. In our case this is:
+
+```python
+def init_fraud_proof(latest_blockhash: bytes, invalid_epoch_commitment: int):
+    ''' Invalid epoch commitment is the index of the invalid blockhash '''
+    # ...
+```
+
+What makes rollups special is that all of the preimages to the `latest_blockhash` are known. This means that from within the fraud proof VM, all blockhash preimages (including the state root, pre blockhash, etc) can be queried using the `preimage oracle`. Fundamentally this means that all historical blocks, events, transactions, and state can are accessible within the fraud proof VM.
+
+### Preimage Oracle
+
+The fraud proof VM has access to the following function (called the "preimage oracle"):
+
+```python
+def get_preimage(hash: bytes) -> bytes:
+    # The preimage is looked up from a local database held by the fraud prover
+    preimage = fetch_preimage(hash)
+    assert keccak256(preimage) == hash
+    return preimage
+```
+
+This seemingly simple functionality has extreme consequences. All publically available preimages can be operated on from within the fraud proof -- a pretty amazing property.
+
+### Step 1: Make available the blockhash disagreement
+- Requirement (1) reminder: the proposer does not submit all intermediate blockhashes.
+- Requirement (2) reminder: The fraud proof VM must only execute a single blockhash.
+
+The only way for us to satisfy these two requirements is to break our fraud proof into two distinct steps. 1) make all the blockhashes that were skipped available, and then 2) use the blockhashes & other Ethereum chain data inside the fraud proof VM to determine the invalid blockhash and execute the transition.
+
+So for step 1 all we need to do is force the proposer to post and commit to all intermediate blockhashes.
```

Comment:
Ahh so in:

> Not all blockhashes are posted up front.

by "up front" what I meant was block hashes are not posted **unless** there is a fraud proof that is being executed. This way in the happy case we don't need to post a whole lot of state roots, but in the sad case we can still retrieve them.

Btw note that this whole thing is not strictly necessary, but limiting the fraud proof to only evaluate a single block vs a range of blocks will save a lot of compute. Maybe if our fraud proof became blazing fast it could be OK though...?

Reply:
The prover will have to prove a bunch of other things anyway, so might be worth keeping the relative simplicity of fraud proofing a single block rather than a range of blocks?

Reply:
Well there is a tradeoff in complexity with proving a single block vs multiple blocks given that we (for efficiency reasons) don't want to post the block hash for every rollup block. The tradeoff is as follows:

- [multi block approach] Proving the range of blocks makes the fraud proof definition extremely trivial. Start from the last block hash that you agree on, and then execute the fraud proof program (playing the dispute game) to determine what the next block hash assertion should have been. 
  - Pro: We only need to run a single fraud proof program - super super easy!
  - Con: It's a lot of computational complexity generating such a large proof!
- [single block approach] In this approach (given we don't post every block hash), we need to have two fraud proof games: 1) determine the block hash we actually disagree on, and then 2) determine what the real block hash should be.
  - Pros: More efficient to generate the proofs (just wayyy less work that you have to do because you don't need to run the fraud proof for most of the blocks)
  - Cons: Now you've got 2 nested dispute games -- kinda sucks :'(

Reply:
I see, I was missing the idea that you didn't want to post every block hash.

"Efficiency reason" = saving on storage costs? Or is it something else?

Wouldn't it be feasible to post every block hash, signed, to calldata? Then a challenge could provide a proof for that bit of calldata, and the signature can be verified, and then challenge game can proceed on that blockhash?